File: kaleidoscope-rs/src/lexer.rs
#[derive(Debug, PartialEq, Clone)]
pub enum Token {
    Eof,
    Def,
    Extern,
    Identifier(String),
    Number(f64),
    LParen,
    RParen,
    Comma,
    Semicolon,
    Other(char),
}

pub fn lex(input: &[char]) -> Vec<Token> {
    let mut idx = 0;
    let mut tokens: Vec<Token> = vec![];
    while idx < input.len() {
        while idx < input.len() && input[idx].is_whitespace() {
            idx += 1;
        }
        if idx >= input.len() {
            break;
        }
        if input[idx].is_alphabetic() {
            let mut id_str = String::new();
            while input[idx].is_alphanumeric() {
                id_str.push(input[idx]);
                idx += 1;
            }
            match id_str.as_str() {
                "def" => tokens.push(Token::Def),
                "extern" => tokens.push(Token::Extern),
                _ => tokens.push(Token::Identifier(id_str)),
            }
        } else if input[idx].is_numeric() || input[idx] == '.' {
            let mut num_str = String::new();
            while input[idx].is_numeric() || input[idx] == '.' {
                num_str.push(input[idx]);
                idx += 1;
            }
            tokens.push(Token::Number(num_str.parse().unwrap()));
        } else if input[idx] == '#' {
            while input[idx] != '\n' && input[idx] != '\r' {
                idx += 1;
            }
        } else {
            let tok = match input[idx] {
                '(' => Token::LParen,
                ')' => Token::RParen,
                ',' => Token::Comma,
                ';' => Token::Semicolon,
                _ => Token::Other(input[idx]),
            };
            tokens.push(tok);
            idx += 1;
        }
    }

    tokens.push(Token::Eof);
    tokens
}

File: kaleidoscope-cpp/src/lexer.cpp
#include "token.hpp"
#include "types.h"
#include <string>
#include <vector>

std::vector<Token> lex(const std::string& input) {
    usize idx = 0;
    std::vector<Token> tokens = {};

    while (idx < input.length()) {
        // Skip whitespace
        while (input[idx] && isspace(input[idx])) {
            idx++;
        }
        if (idx >= input.length()) {
            break;
        }
        if (isalpha(input[idx])) {
            // Lex keyword or identifier
            usize start = idx;
            std::string idStr = "";
            while (isalnum(input[idx])) {
                idStr += input[idx];
                idx++;
            }
            if (idStr == "def") {
                tokens.emplace_back(Token::Tag::Def, start);
            } else if (idStr == "extern") {
                tokens.emplace_back(Token::Tag::Extern, start);
            } else {
                tokens.emplace_back(Token::Tag::Id, start);
            }
        } else if (isdigit(input[idx]) || input[idx] == '.') {
            // Lex number
            usize start = idx;
            while (isdigit(input[idx]) || input[idx] == '.') {
                idx++;
            }
            tokens.emplace_back(Token::Tag::Num, start);
        } else if (input[idx] == '#') {
            // Skip comment
            while (input[idx] && input[idx] != '\n' && input[idx] != '\r') {
                idx++;
            }
        } else {
            switch (input[idx]) {
            case '(':
                tokens.emplace_back(Token::Tag::LParen, idx);
                break;
            case ')':
                tokens.emplace_back(Token::Tag::RParen, idx);
                break;
            case ';':
                tokens.emplace_back(Token::Tag::Semicolon, idx);
                break;
            case '+':
                tokens.emplace_back(Token::Tag::Plus, idx);
                break;
            case '-':
                tokens.emplace_back(Token::Tag::Minus, idx);
                break;
            case '*':
                tokens.emplace_back(Token::Tag::Star, idx);
                break;
            default:
                tokens.emplace_back(Token::Tag::Other, idx);
            }
            idx++;
        }
    }

    tokens.emplace_back(Token::Tag::Eof, idx);
    return tokens;
};
